### ISSUE TITLE
support openssl 1.1.x

### ISSUE THREAD

**Issue by mrbaseman** [adrienverge/openfortivpn#145]
In the long term we should support openssl 1.1

Currently, when linking against openssl 1.1.x,  compilation doesn't succeed and therefore we have added a check in the configure script to preserve users from running configure successful and failing afterwards at compile time

**Comment by mrbaseman** [comment]
I have just seen that openssh is discussing the same topic on their [devel list](https://lists.mindrot.org/pipermail/openssh-unix-dev/2017-June/thread.html) and also OpenVPN has made an effort to [support 1.1](https://www.mail-archive.com/openvpn-devel%40lists.sourceforge.net/msg14946.html). I'm not expecting that approaches can be applied 1:1, but for reference of more general discussions and a first impression of the consequences of the [API changes](https://wiki.openssl.org/index.php/OpenSSL_1.1.0_Changes) the former links could be good resources, the latter one being the official documentation.

**Comment by rahra** [comment]
What exactly is the problem with OpenSSL 1.1?
I just compiled against 1.1.0f and it compiles and the client works.

**Comment by mrbaseman** [comment]
I have seen the following error messages: 
> src/io.c:68:13: error: ‘lock_callback’ defined but not used [-Werror=unused-function]
>  static void lock_callback(int mode, int type, char *file, int line)
>              ^
> src/io.c:75:22: error: ‘thread_id’ defined but not used [-Werror=unused-function]
>  static unsigned long thread_id()
>                       ^
> cc1: all warnings being treated as errors
> Makefile:532: recipe for target 'src/openfortivpn-io.o' failed
> make: *** [src/openfortivpn-io.o] Error 1

and indeed with https://github.com/adrienverge/openfortivpn/commit/3531e2bf29f7dba8af78f1e647a04e63067255d6 they go away and the code at least compiles. 
When I have opened this issue I didn't have a deeper look if it was such a small fix or if larger code areas need changes after the first warnings are fixed.
Also, I haven't tested yet if the connection actually works. It might also be something to test against different versions of FortiOS on the server side.

**Comment by rahra** [comment]
Yes, but these errors are not related to openssl.
It is the code checking option "-Werror" of gcc. These compiler errors occur because the functions are declared "static" and are not called within this file because they are callback functions for openssl.

**Comment by fabiopicchi** [comment]
I have just tried to install openfortivpn at Debian 9. The default version of the OpenSSL packages is greater than 1.1.0, and in the configure script there was an explicit check that I had to remove in order to build openfortivpn. I commented out lines 4636-4652.

Also, when running the Makefile generated, I had to remove the '-Werror' option from the gcc command.

Now, everything is installed and working properly :)

I am sorry if my comment is not relevant to this thread. I reckon that these scripts could be updated so less experienced users do not face these difficulties. I am quite a noob myself, by I can try to look into it if you think it makes sense.

Thanks!
